### PR TITLE
Btc bridge redeem

### PR DESCRIPTION
**Pull request by liujiayu2**
add redeem.rs moudle and import it in lib.rs with "moudle redeem.rs"

### CODE REVIEW

File: xpallets/gateway/bitcoin/v2/src/redeem.rs [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+        ForceRedeemIsAccepted,
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// redeem request id is not exsit
+        RedeemNotExist,
```
Comment (alannotnerd) [inline xpallets/gateway/bitcoin/v2/src/redeem.rs]:
```suggestion
        RedeemRequestNotFound,
```

File: xpallets/gateway/bitcoin/v2/src/redeem.rs [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+
+        /// `RedeemRequest` cancelled for forced redeem when it's not expired.
+        RedeemRequestNotExpired,
+
+        /// Vault is under Liquidation
+        ValtLiquidated,
+
+        /// redeem amount is too much
+        RedeemAmountLargerThanHave,
```
Comment (alannotnerd) [inline xpallets/gateway/bitcoin/v2/src/redeem.rs]:
```suggestion
        InsufficiantAssetsFonds
```

File: xpallets/gateway/bitcoin/v2/src/redeem.rs [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+
+        /// redeem is completed
+        CancleRedeemErrOfCompleted,
```
Comment (alannotnerd) [inline xpallets/gateway/bitcoin/v2/src/redeem.rs]:
```suggestion
        RedeemRequestAlreadyCompleted,
```

File: xpallets/gateway/bitcoin/v2/src/redeem.rs [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+
+        /// redeem is cancled
+        CancleRedeemErrOfCancled,
```
Comment (alannotnerd) [inline xpallets/gateway/bitcoin/v2/src/redeem.rs]:
```suggestion
        RedeemRequestAlreadyCanceled,
```

File: xpallets/gateway/bitcoin/v2/src/redeem.rs [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+            let height = <frame_system::Pallet<T>>::block_number();
+            let vault = vaultpallet::Pallet::<T>::get_active_vault_by_id(&vault_id)?;
+
+            //check current bridge status
+            let bridge_status = <BridgeStatus<T>>::get();
+            match bridge_status {
+                crate::assets::types::Status::Running => {}
+                crate::assets::types::Status::Error => {
+                    Self::deposit_event(Event::<T>::ChainStatusErro);
+                    return Ok(().into());
```
Comment (alannotnerd) [inline xpallets/gateway/bitcoin/v2/src/redeem.rs:166]: Why break out here?

File: xpallets/gateway/bitcoin/v2/src/redeem.rs [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+                }
+            }
+
+            // decrase user's XBTC amount
+
+            // decrese vault's lock collateral.
+            let worth_pcx = assetspallet::Pallet::<T>::convert_to_pcx(redeem_amount)?;
+            assetspallet::Pallet::<T>::lock_collateral(&vault_id, worth_pcx)?;
```
Comment (alannotnerd) [inline xpallets/gateway/bitcoin/v2/src/redeem.rs:178]: why lock again?

File: xpallets/gateway/bitcoin/v2/src/redeem.rs [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+
+            // tell vault to transfer btc to this btc_addr
+
+            // generate redeem request identify
+            let redeemid: super::types::RedeemRequestIdentify = Vec::new();
+            let redeemidentify: super::types::RedeemRequestIdentify = redeemid;
```
Comment (alannotnerd) [inline xpallets/gateway/bitcoin/v2/src/redeem.rs:184]: Thus  all `RedeemRequest` will have same id!

File: xpallets/gateway/bitcoin/v2/src/redeem.rs [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+                    requester: sender,
+                    btc_address: btc_addr,
+                    amount: redeem_amount,
+                    redeem_fee: Default::default(),
+                    status: super::types::RedeemRequestStatus::WaitForGetBtc,
+                },
+            );
+
+            // Redeem id increase
+            <RedeemId<T>>::mutate(|c| *c += 1);
```
Comment (alannotnerd) [inline xpallets/gateway/bitcoin/v2/src/redeem.rs:201]: replace the fake redeemid mentioned with this.

File: xpallets/gateway/bitcoin/v2/src/redeem.rs [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+            ensure!(
+                request.status != super::types::RedeemRequestStatus::Completed,
+                Error::<T>::CancleRedeemErrOfCompleted
+            );
+
+            // catulate user's XBTC worth how much pcx, then give he the pcx
+            let worth_pcx = assetspallet::Pallet::<T>::convert_to_pcx(request.amount)?;
+
+            // add user pcx amount of worth_pcx.then we should sub who's pcx ????
+            assetspallet::Pallet::<T>::slash_collateral(&sender, &sender, worth_pcx);
```
Comment (alannotnerd) [inline xpallets/gateway/bitcoin/v2/src/redeem.rs:267]: from vault to sender

**Review by alannotnerd — CHANGES_REQUESTED** [review]

File: xpallets/gateway/bitcoin/v2/src/redeem.rs (hunk 5 above)
Comment (liujiayu2) [reply xpallets/gateway/bitcoin/v2/src/redeem.rs:166]: my thought is that the bridge has some not correct status at this time and should wait system solve then go on working.

File: xpallets/gateway/bitcoin/v2/src/redeem.rs (hunk 9 above)
Comment (liujiayu2) [reply xpallets/gateway/bitcoin/v2/src/redeem.rs:267]: ok.got it

File: xpallets/gateway/bitcoin/v2/src/redeem.rs (hunk 7 above)
Comment (liujiayu2) [reply xpallets/gateway/bitcoin/v2/src/redeem.rs:184]: ok.I will change this

File: xpallets/gateway/bitcoin/v2/src/redeem.rs (hunk 8 above)
Comment (liujiayu2) [reply xpallets/gateway/bitcoin/v2/src/redeem.rs:201]: ok. got it

File: xpallets/gateway/bitcoin/v2/src/redeem.rs (hunk 6 above)
Comment (liujiayu2) [reply xpallets/gateway/bitcoin/v2/src/redeem.rs:178]: Business understanding error, will be corrected

**Pull request closed without merging**